src/server/streamableHttp.ts

```
@@ -0,0 +1,397 @@
+      const isInitializationRequest = messages.some(
+        msg => 'method' in msg && msg.method === 'initialize'
+      );
+      if (isInitializationRequest) {
+        if (this._initialized) {
+          res.writeHead(400).end(JSON.stringify({
+            jsonrpc: "2.0",
+            error: {
+              code: -32600,
+              message: "Invalid Request: Server already initialized"
+            },
+            id: null
+          }));
+          return;
+        }
+        if (messages.length > 1) {
+          res.writeHead(400).end(JSON.stringify({
+            jsonrpc: "2.0",
+            error: {
+              code: -32600,
+              message: "Invalid Request: Only one initialization request is allowed"
+            },
+            id: null
+          }));
+          return;
+        }
+        this.sessionId = this.sessionIdGenerator();
+        this._initialized = true;
+        const headers: Record<string, string> = {};
+
+        if (this.sessionId !== undefined) {
+          headers["mcp-session-id"] = this.sessionId;
+        }
+
+        // Process initialization messages before responding
+        for (const message of messages) {
+          this.onmessage?.(message);
+        }
+
+        res.writeHead(200, headers).end();
+        return;
+      }
+      // If an Mcp-Session-Id is returned by the server during initialization,
+      // clients using the Streamable HTTP transport MUST include it 
+      // in the Mcp-Session-Id header on all of their subsequent HTTP requests.
+      if (!isInitializationRequest && !this.validateSession(req, res)) {
+        return;
+      }
+
+
+      // check if it contains requests
+      const hasRequests = messages.some(msg => 'method' in msg && 'id' in msg);
+      const hasOnlyNotificationsOrResponses = messages.every(msg =>
+        ('method' in msg && !('id' in msg)) || ('result' in msg || 'error' in msg));
```

Are these ever both going to be false? (Maybe all errors?) If so we will fail to return a response at all unless I'm missing something.

If they are never both going to be false do you need both?

---

> If they are never both going to be false do you need both?

Yes agreed, `hasOnlyNotificationsOrResponses` could probably be inlined into `!hasRequests`.
There should be different validation on the messages based on [the spec](https://modelcontextprotocol.io/specification/2025-03-26/basic/transports#sending-messages-to-the-server) since

> The body of the POST request MUST be one of the following:
> A single JSON-RPC request, notification, or response
> An array [batching](https://www.jsonrpc.org/specification#batch) one or more requests and/or notifications
> An array [batching](https://www.jsonrpc.org/specification#batch) one or more responses

is not correctly enforced as-is.

--- 
> If so we will fail to return a response at all 

Yes - this is to satisfy:

> If the server accepts the input, the server MUST return HTTP status code 202 Accepted with no body.

(should actually probably be reworded to mention "accepts the input when there are no requests" or something, CTTOI)

and accords with [JSON-RPC batch ](https://www.jsonrpc.org/specification#batch) spec

>  A Response object SHOULD exist for each Request object
> ...
> there SHOULD NOT be any Response objects for notifications
> ...
> If there are no Response objects contained within the Response array as it is to be sent to the client, the server MUST NOT return an empty Array and should return nothing at all.

so `res.writeHead(202).end();` below intentionally responds immediately with no body, the server can process the messages without blocking the ServerResponse.

This should be clarified with some additional code comments, imo.